### ISSUE TITLE
display sub millisecond

### ISSUE THREAD

**Issue by amdkam** [sqlitebrowser/sqlitebrowser#3368]
First of all, thanks for the great tool you've built, it works like a charm.

One issue I run into is working with timestamps at sub millisecond precision.
I have a field called 'Time' REAL with timestamps that are displayed as: 0.603871998 
I would like to have these formatted as HH:MM:SS.SSSSSS to keep microsecond resolution.

According to https://www.sqlite.org/lang_datefunc.html#overview  I should be able to achieve this using strftime() and the 'subsec' or 'subsecond' modifier, but this will render all values in 'Time' as NULL. 

The only query I found somewhat usable is the following: 
`SELECT strftime('%H:%M:%f', Time,'unixepoch') AS FormattedTime`
Time in this case is a timestamp like: 0.603871998  which will be displayed as: 00:00:00.604 
so 3 digits max.

Please let me know if I'm missing something here, otherwise I would like to request this feature.

DB Browser for SQLite Version 3.12.2

Built for x86_64-little_endian-llp64, running on x86_64
Qt Version 5.12.8
SQLite Version 3.35.5.

**Comment by i-s-o** [comment]
**TLDR:**
If your "Time" field represents seconds, here are two simple solutions:
* Standard functions:
  ```sql
  select format('%02.f', Time/3600) || ':' || format('%02.f', mod(Time, 3600)/60) || ':' || format('%012.9f', mod(Time, 60)) as FormattedTime;
  ```
* If millisecond precision is good enough, you can use a shorter formula using date/time functions:
  ```sql
  select strftime('%H:%M:%f', Time / 86400.0 + 0.5);
  ```

**Explanation of the second option:**
SQLite doesn't have a built-in Date, Time, or DateTime data types. Instead, according to [SQLite docs](https://www.sqlite.org/lang_datefunc.html), you need to store them in one of the three built-in types to perform further date/time operations on them:
1. text in a subset of the [ISO-8601](http://en.wikipedia.org/wiki/ISO_8601) format,
2. numbers representing the [Julian day](http://en.wikipedia.org/wiki/Julian_day), or
3. numbers representing the number of seconds since (or before) 1970-01-01 00:00:00 UTC (the unix timestamp)

In your situation, a value like 0.603871998 in your "Time" field fits the option 2 above. This is because option 1 expects STRING, and option 3 expects INTEGER, so, SQLite would interpret 0.603871998 as the Julian day value if you use it in date/time functions. And, according to the Wikipedia article linked above, that value represents 02:29:34.541 a.m. on November 25, 4713 BC in the [proleptic Gregorian calendar](https://en.wikipedia.org/wiki/Proleptic_Gregorian_calendar). You can also see this in SQLite:
```sql
select strftime('%Y-%m-%d %H:%M:%f', 0.603871998); --> -4713-11-25 02:29:34.541
```

The second solution above converts "Time" (assuming it is in seconds) to a fraction of the day so that SQLite can interpret it as a DateTime value, then, '%H:%M:%f' format string simply ignores the date part. Also, since the Julian day starts at noon we need to add half a day (0.5) to count from zero (i.e. midnight).
```sql
select strftime('%Y-%m-%d %H:%M:%f', 0.603871998 / 86400.0); --> -4713-11-24 12:00:00.604
select strftime('%Y-%m-%d %H:%M:%f', 0.603871998 / 86400.0 + 0.5); --> -4713-11-25 00:00:00.604
select strftime('%H:%M:%f', 0.603871998 / 86400.5 + 0.5); --> 00:00:00.604
```

**Comment by amdkam** [comment]
Thanks for the helpful response.
The first option works for me, I get the microsecond precision, perfect. Though it only works on the nightly build (3.12.99), because both format() and mod() are not implemented yet in the latest stable build (3.12.2). But for now this solved the issue.

n.b. Apparently I overlooked another line in the sqlite docs:
"The current implemention increases the resolution from seconds to milliseconds, but this might increase to a higher resolution in future releases of SQLite. "

**Comment by chrisjlocke** [comment]
Closing as completed

**Comment by amdkam** [comment]
Came across this extension and figured it might contribute to this discussion.
https://antonz.org/sqlean-time/
> The sqlean-time extension provides functionality for working with time and duration with nanosecond precision.

**Comment by lucydodo** [comment]
Hi @amdkam,
Thank you for not forgetting about this issue and sharing that information with us. :)

Actually, we already have that extension built-in, but for some reason I've noticed that the newly introduced functions don't work. I'll fix it in the upcoming 3.13.1 release, and thanks again for letting us know.

**Comment by lucydodo** [comment]
I've checked, and it's caused by the upstream side not adding new extensions to the unified extension file.
I've opened an issue upstream to see if this is intentional behavior or not.

**Comment by lucydodo** [comment]
The upstream side has fixed the issue, so our app should too.
The next nightly build process is scheduled in about 8 hours, so we'll check it out then and let you know.

**Comment by lucydodo** [comment]
I've checked that it's working fine in the current nightly build, if you don't mind, could you give it a try?
You can download the nightly build [here](https://nightlies.sqlitebrowser.org/latest/).
